precise/iaf_psc_alpha_ps.cpp

```diff
@@ -500,37 +503,32 @@ nest::iaf_psc_alpha_ps::handle( DataLoggingRequest& e )
 void
 nest::iaf_psc_alpha_ps::propagate_( const double dt )
 {
-  // needed in any case
-  const double ps_e_TauSyn_ex = numerics::expm1( -dt / P_.tau_syn_ex_ );
-  const double ps_e_TauSyn_in = numerics::expm1( -dt / P_.tau_syn_in_ );
-
   // V_m_ remains unchanged at 0.0 while neuron is refractory
   if ( not S_.is_refractory_ )
   {
-    const double ps_e_Tau = numerics::expm1( -dt / P_.tau_m_ );
-    const double ps_P30 = -P_.tau_m_ / P_.c_m_ * ps_e_Tau;
-
-    const double ps_P31_ex = V_.gamma_sq_ex_ * ps_e_Tau - V_.gamma_sq_ex_ * ps_e_TauSyn_ex
-      - dt * V_.gamma_ex_ * ps_e_TauSyn_ex - dt * V_.gamma_ex_;
-    const double ps_P32_ex = V_.gamma_ex_ * ps_e_Tau - V_.gamma_ex_ * ps_e_TauSyn_ex;
+    const double ps_P30 = -P_.tau_m_ / P_.c_m_ * numerics::expm1( -dt / P_.tau_m_ );
 
-    const double ps_P31_in = V_.gamma_sq_in_ * ps_e_Tau - V_.gamma_sq_in_ * ps_e_TauSyn_in
-      - dt * V_.gamma_in_ * ps_e_TauSyn_in - dt * V_.gamma_in_;
-    const double ps_P32_in = V_.gamma_in_ * ps_e_Tau - V_.gamma_in_ * ps_e_TauSyn_in;
+    const double ps_P31_ex = propagator_31( P_.tau_syn_ex_, P_.tau_m_, P_.c_m_, dt );
+    const double ps_P32_ex = propagator_32( P_.tau_syn_ex_, P_.tau_m_, P_.c_m_, dt );
+    const double ps_P31_in = propagator_31( P_.tau_syn_in_, P_.tau_m_, P_.c_m_, dt );
+    const double ps_P32_in = propagator_32( P_.tau_syn_in_, P_.tau_m_, P_.c_m_, dt );
 
     S_.V_m_ = ps_P30 * ( P_.I_e_ + S_.y_input_ ) + ps_P31_ex * S_.dI_ex_ + ps_P32_ex * S_.I_ex_ + ps_P31_in * S_.dI_in_
-      + ps_P32_in * S_.I_in_ + ps_e_Tau * S_.V_m_ + S_.V_m_;
+      + ps_P32_in * S_.I_in_ + S_.V_m_ * std::exp( -dt / P_.tau_m_ );
```

Comment:
Can we rewrite this expression to its equivalent expm1 form for better floating-point precision?

Reply:
We have this equation essentially in three places: lines 364f, 516f and 588ff. We should use the same implementation everywhere. In the `test_iaf_psc_dc...accurary.sli` tests, the `V * expm1() + V` curiously leads to smaller errors than `V * exp()` in the presence of pure DC input; so it may be the better choice.